File: Examples/Advanced/Export-ThirdEye-Violations/nodeJS/export-thirdeye-violations.mjs
#!/usr/bin/env node

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

export const STATE_FORMAT = "logicvein-thirdeye-violation-export-state";
export const RUN_FORMAT = "logicvein-thirdeye-violation-export-run";
const FORMAT_VERSION = 1;
export const CSV_FIELDS = [
  "eventId", "incidentId", "severity", "clearState", "eventType", "network",
  "ipAddress", "hostname", "deviceId", "hostUuid", "measurement",
  "measurementIndex", "message", "occurrences", "triggerId", "created", "updated",
];

export class ExampleError extends Error {}

export function loadEnv(file) {
  if (!fs.existsSync(file)) return;
  for (const raw of fs.readFileSync(file, "utf8").split(/\r?\n/)) {
    const match = raw.match(/^\s*(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)\s*$/);
    if (!match || raw.trimStart().startsWith("#")) continue;
    let value = match[2];
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }
    process.env[match[1]] = value;
  }
}

export function parseQueries(value) {
  let parsed;
  try {
    parsed = JSON.parse(value || "[]");
  } catch (error) {
    throw new ExampleError("NETLD_SEARCH_QUERIES must be a JSON array of strings.", { cause: error });
  }
  if (!Array.isArray(parsed) || parsed.some((item) => typeof item !== "string" || !item.trim())) {
    throw new ExampleError("NETLD_SEARCH_QUERIES must be a JSON array of non-empty strings.");
  }
  const queries = parsed.map((item) => item.trim());
  if (queries.some((item) => /^(?:start|end)=/i.test(item))) {
    throw new ExampleError("NETLD_SEARCH_QUERIES must not contain start or end; the exporter controls its time window.");
  }
  return queries;
}

const resolvePath = (base, value) => path.isAbsolute(value) ? value : path.join(base, value);
const isoUtc = (milliseconds) => new Date(Number(milliseconds)).toISOString().replace(/\.\d{3}Z$/, "Z");

function positiveInteger(value, name) {
  const result = Number(value);
  if (!Number.isInteger(result) || result <= 0) throw new ExampleError(`${name} must be a positive integer.`);
  return result;
}

export function configFromEnv(base) {
  for (const name of ["NETLD_BASE_URL", "NETLD_API_KEY"]) {
    if (!process.env[name]?.trim()) throw new ExampleError(`Set ${name} in the environment file.`);
  }
  const outputFormat = (process.env.NETLD_OUTPUT_FORMAT || "csv").trim().toLowerCase();
  if (!["csv", "json"].includes(outputFormat)) throw new ExampleError("NETLD_OUTPUT_FORMAT must be csv or json.");
  return {
    baseUrl: process.env.NETLD_BASE_URL.trim().replace(/\/$/, ""),
    apiKey: process.env.NETLD_API_KEY.trim(),
    outputDir: resolvePath(base, process.env.NETLD_OUTPUT_DIR || "violation-exports"),
    outputFormat,
    statePath: resolvePath(base, process.env.NETLD_STATE_FILE || "violation-export-state.json"),
    reportPath: resolvePath(base, process.env.NETLD_RUN_REPORT_FILE || "violation-export-run.json"),
    pageSize: positiveInteger(process.env.NETLD_PAGE_SIZE || "100", "NETLD_PAGE_SIZE"),
    initialLookbackHours: positiveInteger(
      process.env.NETLD_INITIAL_LOOKBACK_HOURS || "24",
      "NETLD_INITIAL_LOOKBACK_HOURS",
    ),
    searchQueries: parseQueries(process.env.NETLD_SEARCH_QUERIES || "[]"),
  };
}

export class ThirdEyeClient {
  constructor(config) {
    this.config = config;
    this.cookie = "";
  }

  async request(endpoint, options = {}) {
    const headers = {
      Authorization: `Bearer ${this.config.apiKey}`,
      ...(this.cookie ? { Cookie: this.cookie } : {}),
      ...(options.headers || {}),
    };
    const response = await fetch(`${this.config.baseUrl}${endpoint}`, {
      ...options,
      headers,
      redirect: "manual",
      signal: AbortSignal.timeout(30000),
    });
    const cookie = response.headers.get("set-cookie");
    if (cookie) this.cookie = cookie.split(";", 1)[0];
    if (response.status >= 300 && response.status < 400) {
      throw new ExampleError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new ExampleError(`Request failed with HTTP ${response.status}.`);
    return response;
  }

  async login() {
    await this.request("/rest");
  }

  async searchPage(queries, offset, pageSize) {
    const response = await this.request("/rest", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        method: "Incidents.searchTriggerEvents",
        params: {
          pageData: { offset, total: 0, pageSize, violations: [] },
          queries,
          sortColumn: "updated",
          descending: true,
        },
        id: crypto.randomUUID(),
      }),
    });
    const data = await response.json();
    if (data.error) throw new ExampleError(`Incidents.searchTriggerEvents failed: ${JSON.stringify(data.error)}`);
    if (!data.result || typeof data.result !== "object" || !Array.isArray(data.result.violations)) {
      throw new ExampleError("Incidents.searchTriggerEvents returned invalid page data.");
    }
    return data.result;
  }
}

function numericField(event, field) {
  const value = Number(event?.[field]);
  if (!Number.isInteger(value)) {
    throw new ExampleError(`Incidents.searchTriggerEvents returned an invalid ${field}.`);
  }
  return value;
}

const eventId = (event) => numericField(event, "eventId");
const eventUpdated = (event) => numericField(event, "updated");

function loadState(file) {
  if (!fs.existsSync(file)) return { format: STATE_FORMAT, formatVersion: FORMAT_VERSION };
  const state = JSON.parse(fs.readFileSync(file, "utf8"));
  if (state.format !== STATE_FORMAT || state.formatVersion !== FORMAT_VERSION) {
    throw new ExampleError("The violation-export state file has an unsupported format.");
  }
  return state;
}

function atomicWrite(file, content) {
  fs.mkdirSync(path.dirname(file), { recursive: true });
  const temporary = path.join(path.dirname(file), `.${path.basename(file)}.${crypto.randomUUID()}.tmp`);
  try {
    fs.writeFileSync(temporary, content, { mode: 0o600 });
    fs.renameSync(temporary, file);
  } catch (error) {
    try { fs.unlinkSync(temporary); } catch {}
    throw error;
  }
}

const writeJson = (file, value) => atomicWrite(file, `${JSON.stringify(value, null, 2)}\n`);

export function buildQueries(config, state, nowMs) {
  const start = state.lastUpdated == null
    ? nowMs - config.initialLookbackHours * 60 * 60 * 1000
    : Number(state.lastUpdated);
  return [...config.searchQueries, `start=${isoUtc(start)}`, `end=${isoUtc(nowMs)}`];
}

export async function collectEvents(client, queries, pageSize) {
  let offset = 0;
  let total = null;
  let pages = 0;
  const events = [];
  while (total == null || offset < total) {
    const page = await client.searchPage(queries, offset, pageSize);
    const batch = page.violations;
    const pageOffset = Number(page.offset ?? offset);
    const pageTotal = Number(page.total ?? pageOffset + batch.length);
    total = total == null ? pageTotal : Math.max(total, pageTotal);
    pages += 1;
    for (const event of batch) {
      if (!event || typeof event !== "object" || Array.isArray(event)) {
        throw new ExampleError("Incidents.searchTriggerEvents returned a non-object violation.");
      }
      eventId(event);
      eventUpdated(event);
      events.push(event);
    }
    const nextOffset = pageOffset + batch.length;
    if (nextOffset >= total) break;
    if (!batch.length || nextOffset <= offset) {
      throw new ExampleError("Violation paging stopped before all reported results were returned.");
    }
    offset = nextOffset;
  }
  const unique = new Map(events.map((event) => [`${eventUpdated(event)}:${eventId(event)}`, event]));
  return {
    events: [...unique.values()].sort((left, right) => eventUpdated(left) - eventUpdated(right) || eventId(left) - eventId(right)),
    pageCount: pages,
  };
}

export function selectEvents(events, state) {
  if (state.lastUpdated == null) return events;
  const watermark = Number(state.lastUpdated);
  const ids = new Set((state.eventIdsAtLastUpdated || []).map(Number));
  return events.filter((event) => eventUpdated(event) > watermark
    || (eventUpdated(event) === watermark && !ids.has(eventId(event))));
}

function csvCell(value) {
  const text = value == null ? "" : String(value);
  return /[",\r\n]/.test(text) ? `"${text.replaceAll('"', '""')}"` : text;
}

export function renderExport(events, outputFormat) {
  if (outputFormat === "json") return `${JSON.stringify(events, null, 2)}\n`;
  const rows = [CSV_FIELDS.join(",")];
  for (const event of events) {
    rows.push(CSV_FIELDS.map((field) => {
      const value = ["created", "updated"].includes(field) && event[field] != null
        ? isoUtc(event[field])
        : event[field];
      return csvCell(value);
    }).join(","));
  }
  return `${rows.join("\n")}\n`;
}

function advanceState(state, events) {
  if (!events.length) return;
  const newest = Math.max(...events.map(eventUpdated));
  const ids = events.filter((event) => eventUpdated(event) === newest).map(eventId);
  if (state.lastUpdated != null && Number(state.lastUpdated) === newest) {
    ids.push(...(state.eventIdsAtLastUpdated || []).map(Number));
  }
  state.lastUpdated = newest;
  state.eventIdsAtLastUpdated = [...new Set(ids)].sort((left, right) => left - right);
}

export async function exportViolations(client, config, nowMs, timestamp = isoUtc(nowMs)) {
  const state = loadState(config.statePath);
  const queries = buildQueries(config, state, nowMs);
  const { events, pageCount } = await collectEvents(client, queries, config.pageSize);
  const selected = selectEvents(events, state);
  let outputFile = null;
  if (selected.length) {
    const stamp = isoUtc(nowMs).replace(/[-:]/g, "");
    outputFile = path.join(config.outputDir, `violations-${stamp}.${config.outputFormat}`);
    atomicWrite(outputFile, renderExport(selected, config.outputFormat));
    advanceState(state, selected);
  }
  const report = {
    format: RUN_FORMAT,
    formatVersion: FORMAT_VERSION,
    generatedAt: timestamp,
    queries,
    pageCount,
    resultCount: events.length,
    exportedCount: selected.length,
    outputFormat: config.outputFormat,
    outputFile,
  };
  writeJson(config.statePath, state);
  writeJson(config.reportPath, report);
  return report;
}

async function main() {
  const here = path.dirname(fileURLToPath(import.meta.url));
  const index = process.argv.indexOf("--env");
  const envFile = index >= 0 ? process.argv[index + 1] : path.join(here, ".env");
  if (index >= 0 && !envFile) throw new ExampleError("--env requires a file path.");
  loadEnv(envFile);
  const config = configFromEnv(here);
  const client = new ThirdEyeClient(config);
  await client.login();
  const report = await exportViolations(client, config, Date.now());
  console.log(`Found ${report.resultCount} violations and exported ${report.exportedCount}.`);
  console.log(`Run report: ${config.reportPath}`);
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  main().catch((error) => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
}

File: Examples/Advanced/Export-ThirdEye-Violations/Python/export_thirdeye_violations.py
#!/usr/bin/env python3
"""Incrementally export ThirdEye trigger-event violations to CSV or JSON."""

from __future__ import annotations

import argparse
import csv
import io
import json
import os
import sys
import tempfile
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import requests
from dotenv import load_dotenv


STATE_FORMAT = "logicvein-thirdeye-violation-export-state"
RUN_FORMAT = "logicvein-thirdeye-violation-export-run"
FORMAT_VERSION = 1
CSV_FIELDS: tuple[str, ...] = (
    "eventId", "incidentId", "severity", "clearState", "eventType", "network",
    "ipAddress", "hostname", "deviceId", "hostUuid", "measurement",
    "measurementIndex", "message", "occurrences", "triggerId", "created", "updated",
)


class ExampleError(RuntimeError):
    pass


def parse_queries(value: str) -> list[str]:
    try:
        parsed = json.loads(value or "[]")
    except json.JSONDecodeError as exc:
        raise ExampleError("NETLD_SEARCH_QUERIES must be a JSON array of strings.") from exc
    if not isinstance(parsed, list) or any(
        not isinstance(item, str) or not item.strip() for item in parsed
    ):
        raise ExampleError("NETLD_SEARCH_QUERIES must be a JSON array of non-empty strings.")
    queries = [item.strip() for item in parsed]
    if any(item.lower().startswith(("start=", "end=")) for item in queries):
        raise ExampleError(
            "NETLD_SEARCH_QUERIES must not contain start or end; the exporter controls its time window."
        )
    return queries


def iso_utc(milliseconds: int) -> str:
    return (
        datetime.fromtimestamp(milliseconds / 1000, timezone.utc)
        .replace(microsecond=0)
        .isoformat()
        .replace("+00:00", "Z")
    )


def event_id(event: dict[str, Any]) -> int:
    value = event.get("eventId")
    if isinstance(value, bool) or not isinstance(value, (int, float, str)):
        raise ExampleError("Incidents.searchTriggerEvents returned an event without an eventId.")
    try:
        return int(value)
    except (TypeError, ValueError) as exc:
        raise ExampleError("Incidents.searchTriggerEvents returned an invalid eventId.") from exc


def event_updated(event: dict[str, Any]) -> int:
    value = event.get("updated")
    if isinstance(value, bool) or not isinstance(value, (int, float, str)):
        raise ExampleError(
            "Incidents.searchTriggerEvents returned an event without an updated timestamp."
        )
    try:
        return int(value)
    except (TypeError, ValueError) as exc:
        raise ExampleError(
            "Incidents.searchTriggerEvents returned an invalid updated timestamp."
        ) from exc


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    output_dir: Path
    output_format: str
    state_path: Path
    report_path: Path
    page_size: int
    initial_lookback_hours: int
    search_queries: list[str]

    @classmethod
    def from_env(cls, env_path: Path, base: Path) -> "Config":
        load_dotenv(env_path, override=True)
        url = os.getenv("NETLD_BASE_URL", "").strip()
        key = os.getenv("NETLD_API_KEY", "").strip()
        if not url or not key:
            raise ExampleError(
                "Set NETLD_BASE_URL and NETLD_API_KEY in the environment file."
            )

        def destination(name: str, default: str) -> Path:
            value = Path(os.getenv(name, default))
            return value if value.is_absolute() else base / value

        output_format = os.getenv("NETLD_OUTPUT_FORMAT", "csv").strip().lower()
        if output_format not in {"csv", "json"}:
            raise ExampleError("NETLD_OUTPUT_FORMAT must be csv or json.")
        try:
            page_size = int(os.getenv("NETLD_PAGE_SIZE", "100"))
            lookback = int(os.getenv("NETLD_INITIAL_LOOKBACK_HOURS", "24"))
        except ValueError as exc:
            raise ExampleError(
                "NETLD_PAGE_SIZE and NETLD_INITIAL_LOOKBACK_HOURS must be positive integers."
            ) from exc
        if page_size <= 0 or lookback <= 0:
            raise ExampleError(
                "NETLD_PAGE_SIZE and NETLD_INITIAL_LOOKBACK_HOURS must be positive integers."
            )
        return cls(
            url.rstrip("/"),
            key,
            destination("NETLD_OUTPUT_DIR", "violation-exports"),
            output_format,
            destination("NETLD_STATE_FILE", "violation-export-state.json"),
            destination("NETLD_RUN_REPORT_FILE", "violation-export-run.json"),
            page_size,
            lookback,
            parse_queries(os.getenv("NETLD_SEARCH_QUERIES", "[]")),
        )


class ThirdEyeClient:
    def __init__(self, config: Config, timeout: float = 30):
        self.config = config
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {config.api_key}"})

    def request(self, method: str, endpoint: str, **kwargs: Any):
        try:
            response = self.session.request(
                method,
                f"{self.config.base_url}{endpoint}",
                timeout=self.timeout,
                allow_redirects=False,
                **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.config.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(
                f"Request redirected to {response.headers.get('Location', '')}."
            )
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(
                f"Request failed with HTTP {response.status_code}."
            ) from exc
        return response

    def login(self) -> None:
        self.request("GET", "/rest")

    def search_page(
        self, queries: list[str], offset: int, page_size: int
    ) -> dict[str, Any]:
        response = self.request(
            "POST",
            "/rest",
            headers={"Content-Type": "application/json"},
            json={
                "jsonrpc": "2.0",
                "method": "Incidents.searchTriggerEvents",
                "params": {
                    "pageData": {
                        "offset": offset,
                        "total": 0,
                        "pageSize": page_size,
                        "violations": [],
                    },
                    "queries": queries,
                    "sortColumn": "updated",
                    "descending": True,
                },
                "id": str(uuid.uuid4()),
            },
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(
                "Incidents.searchTriggerEvents returned invalid JSON."
            ) from exc
        if data.get("error"):
            raise ExampleError(
                f"Incidents.searchTriggerEvents failed: {json.dumps(data['error'])}"
            )
        result = data.get("result")
        if not isinstance(result, dict) or not isinstance(
            result.get("violations"), list
        ):
            raise ExampleError(
                "Incidents.searchTriggerEvents returned invalid page data."
            )
        return result


def load_state(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {"format": STATE_FORMAT, "formatVersion": FORMAT_VERSION}
    try:
        state = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as exc:
        raise ExampleError(f"Could not read export state: {path}") from exc
    if (
        state.get("format") != STATE_FORMAT
        or state.get("formatVersion") != FORMAT_VERSION
    ):
        raise ExampleError("The violation-export state file has an unsupported format.")
    return state


def write_atomic(path: Path, content: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            dir=path.parent,
            prefix=f".{path.name}.",
            suffix=".tmp",
            delete=False,
        ) as stream:
            temporary = Path(stream.name)
            stream.write(content)
        os.replace(temporary, path)
    except Exception:
        if temporary:
            temporary.unlink(missing_ok=True)
        raise


def write_json(path: Path, value: Any) -> None:
    write_atomic(
        path, (json.dumps(value, indent=2, ensure_ascii=False) + "\n").encode()
    )


def build_queries(config: Config, state: dict[str, Any], now_ms: int) -> list[str]:
    start = (
        int(state["lastUpdated"])
        if state.get("lastUpdated") is not None
        else now_ms - config.initial_lookback_hours * 60 * 60 * 1000
    )
    return [
        *config.search_queries,
        f"start={iso_utc(start)}",
        f"end={iso_utc(now_ms)}",
    ]


def collect_events(
    client: Any, queries: list[str], page_size: int
) -> tuple[list[dict[str, Any]], int]:
    offset = 0
    total: int | None = None
    pages = 0
    events: list[dict[str, Any]] = []
    while total is None or offset < total:
        page = client.search_page(queries, offset, page_size)
        batch = page["violations"]
        page_offset = int(page.get("offset", offset))
        page_total = int(page.get("total", page_offset + len(batch)))
        total = page_total if total is None else max(total, page_total)
        pages += 1
        for event in batch:
            if not isinstance(event, dict):
                raise ExampleError(
                    "Incidents.searchTriggerEvents returned a non-object violation."
                )
            event_id(event)
            event_updated(event)
            events.append(event)
        next_offset = page_offset + len(batch)
        if next_offset >= total:
            break
        if not batch or next_offset <= offset:
            raise ExampleError(
                "Violation paging stopped before all reported results were returned."
            )
        offset = next_offset
    unique = {(event_updated(item), event_id(item)): item for item in events}
    return sorted(
        unique.values(), key=lambda item: (event_updated(item), event_id(item))
    ), pages


def select_events(
    events: list[dict[str, Any]], state: dict[str, Any]
) -> list[dict[str, Any]]:
    if state.get("lastUpdated") is None:
        return events
    watermark = int(state["lastUpdated"])
    ids = {int(value) for value in state.get("eventIdsAtLastUpdated", [])}
    return [
        event
        for event in events
        if event_updated(event) > watermark
        or (event_updated(event) == watermark and event_id(event) not in ids)
    ]


def csv_value(field: str, value: Any) -> Any:
    if field in {"created", "updated"} and value is not None:
        return iso_utc(int(value))
    return "" if value is None else value


def render_export(events: list[dict[str, Any]], output_format: str) -> bytes:
    if output_format == "json":
        return (json.dumps(events, indent=2, ensure_ascii=False) + "\n").encode()
    stream = io.StringIO(newline="")
    writer = csv.DictWriter(
        stream, fieldnames=CSV_FIELDS, extrasaction="ignore", lineterminator="\n"
    )
    writer.writeheader()
    for event in events:
        writer.writerow(
            {field: csv_value(field, event.get(field)) for field in CSV_FIELDS}
        )
    return stream.getvalue().encode("utf-8")


def advance_state(state: dict[str, Any], events: list[dict[str, Any]]) -> None:
    if not events:
        return
    newest = max(event_updated(event) for event in events)
    ids = {event_id(event) for event in events if event_updated(event) == newest}
    if state.get("lastUpdated") is not None and int(state["lastUpdated"]) == newest:
        ids.update(int(value) for value in state.get("eventIdsAtLastUpdated", []))
    state.update(lastUpdated=newest, eventIdsAtLastUpdated=sorted(ids))


def export_violations(
    client: Any, config: Config, now_ms: int, timestamp: str | None = None
) -> dict[str, Any]:
    state = load_state(config.state_path)
    queries = build_queries(config, state, now_ms)
    events, page_count = collect_events(client, queries, config.page_size)
    selected = select_events(events, state)
    output_path: Path | None = None
    if selected:
        stamp = datetime.fromtimestamp(now_ms / 1000, timezone.utc).strftime(
            "%Y%m%dT%H%M%SZ"
        )
        output_path = config.output_dir / (
            f"violations-{stamp}.{config.output_format}"
        )
        write_atomic(output_path, render_export(selected, config.output_format))
        advance_state(state, selected)
    report = {
        "format": RUN_FORMAT,
        "formatVersion": FORMAT_VERSION,
        "generatedAt": timestamp or iso_utc(now_ms),
        "queries": queries,
        "pageCount": page_count,
        "resultCount": len(events),
        "exportedCount": len(selected),
        "outputFormat": config.output_format,
        "outputFile": str(output_path) if output_path else None,
    }
    write_json(config.state_path, state)
    write_json(config.report_path, report)
    return report


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--env", type=Path, default=Path(__file__).with_name(".env"))
    args = parser.parse_args()
    try:
        config = Config.from_env(args.env, Path(__file__).resolve().parent)
        client = ThirdEyeClient(config)
        client.login()
        now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
        report = export_violations(client, config, now_ms)
        print(
            f"Found {report['resultCount']} violations and exported {report['exportedCount']}."
        )
        print(f"Run report: {config.report_path}")
        return 0
    except (ExampleError, OSError, ValueError, KeyError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
